src/channel-qvm.lisp

```diff
@@ -103,55 +91,27 @@
-  (dolist (kraus-map kraus-maps)
-    (apply-kraus-map qvm instr kraus-map)))
-
-(defmethod apply-kraus-map ((qvm channel-qvm) (instr quil:gate-application) kraus-ops)
-  ;; Uniformly at random select one of the kraus operators in KRAUS-OPS to apply
-  ;; for the transition.
-  (let* ((amps (%trial-amplitudes qvm))
-         (r (random 1.0d0))
-         (summed 0.0d0)
-         (logical-qubits (quil:application-arguments instr))
-         (qubits (mapcar #'quil:qubit-index logical-qubits)))
-    (declare (type quantum-state amps))
-    ;; Randomly select one of the Kraus operators by inverse transform
-    ;; sampling (cf [1]): We divide the unit interval [0,1] into n
-    ;; bins where the j-th bin size equals the probability p_j with
-    ;; which the j-th Kraus operator k_j should be applied. We know
-    ;; that the Kraus operators are normalized such that p_j =
-    ;; <psi|k_j^H k_j |psi> where x^H denotes hermitian conjugation of
-    ;; x and can therefore perform this sampling lazily: First
-    ;; generate a uniformly sampled random number r in [0,1]. Next,
-    ;; find j such that
-    ;;
-    ;;       sum_{k=1}^{j-1} p_k < r <= sum_{k=1}^{j} p_k
-    ;;
-    ;; This is possible by evaluating only all p_k for k<=j. Then
-    ;; pick this j as the choice of Kraus operator to apply.
-    ;;
-    ;; [1]: https://en.wikipedia.org/wiki/Inverse_transform_sampling
-    (loop :for kj :in kraus-ops
-          :do
-             (replace amps (amplitudes qvm))
-             (apply-matrix-operator (magicl-matrix-to-quantum-operator kj)
-                                    amps
-                                    (apply #'nat-tuple qubits))
-             (incf summed (psum #'probability amps))
-          :until (>= summed r))
-    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))
-    (normalize-wavefunction (amplitudes qvm))))
-
-(defmethod apply-classical-readout-noise ((qvm channel-qvm) (instr quil:measure-discard))
-  (declare (ignore qvm instr))
-  nil)
-
-(defmethod apply-classical-readout-noise ((qvm channel-qvm) (instr quil:measure))
-  (%corrupt-qvm-memory-with-povm qvm instr (readout-povms (noise-model qvm))))
-
-(defmethod apply-classical-readout-noise ((qvm channel-qvm) (instr compiled-measurement))
-  (apply-classical-readout-noise qvm (source-instruction instr)))
+  ;; Apply each kraus map in the list KRAUS-MAPS to the state of the system, using the qubits from the current instruction INSTR.
+  (let ((qubits (mapcar #'quil:qubit-index (quil:application-arguments instr))))
+    (dolist (kraus-map kraus-maps)
+      (apply-noise-to-state kraus-map (state qvm) qubits))))
+
+(defgeneric apply-classical-readout-noise (qvm instr)
+  (:documentation "Given a QVM and a (measurement) instruction INSTR, corrupt the readout bit according to the POVM specifications of QVM.")
+  ;; Make sure readout noise is never applied to a pure-state qvm
+  (:method ((qvm pure-state-qvm) (instr quil:measurement))
+    (declare (ignore qvm instr))
```

Comment:
Should we signal a condition here?

Reply:
I think so; it seems like a bug if it is.

Alternatively, don't implement the method.